### PR TITLE
Core: test_var is using `exec` instead of calling `var` directly

### LINKED ISSUE #17547: test_var should use `exec` instead of calling `var` directly.
test_var should also probably be using `exec` instead of calling `var` directly. The problem is that once you run the test once, the variables from var are defined globally (`var` injects names into `globals()`, not `locals()`, which is one reason why its use noninteractively is so bad). So the tests could be subtly wrong. For instance, if someone accidentally wrote

```py
def test1():
    var('x')
    assert x == Symbol('x')

def test2():
    var('xx')
    assert x == Symbol('x') # Typo: this should be assert xx == Symbol('xx')
```

the test would pass, but only if test1 is run before test2. It also could lead to unexpected behavior if someone were to run `test(subprocess=False)` twice in the same process (e.g., https://github.com/sympy/sympy/pull/15163). Generally, tests should be written in a way that they do not influence some global state so that they affect each other. It would be better to have tests in that file like

```py
def test_var():
    ns = {'var': var}
    eval("var('x')", ns)
    assert ns['x'] == Symbol('x')
```

_Originally posted by @asmeurer in https://github.com/sympy/sympy/pull/17544#issuecomment-526375244_

### PR DESCRIPTION
**Pull request by sachin-4099**
Fixes#17547

test_var should use 'exec' instead of calling 'var' directly. 
The problem is that once you run the test once, the variables from var are defined globally             
(var injects names into globals(), not locals(), which is one reason why its use non-interactively is so bad). 
So the tests could be subtly wrong.

* core
  * test_var is using `exec` instead of calling `var` directly

### CODE REVIEW

File: sympy/core/tests/test_var.py [hunk 1]
```diff
@@ -1,83 +1,91 @@
...
-    var("z1")
 
 # make z2 with call-depth = 2
 
 
-def __make_z2():
-    var("z2")
-
+def test_var():
+    exec(textwrap.dedent("""
```
Comment (Oktosha) [inline sympy/core/tests/test_var.py]: `textwrap.dedent` removes extra indentation. So, I guess, it's not needed here as you don't add extra indent to the code inside the `exec`

File: sympy/core/tests/test_var.py [hunk 2]
```diff
@@ -1,83 +1,91 @@
 # Tests for var are in their own file, because var pollutes global namespace.
-
+import textwrap
 from sympy import Symbol, var, Function, FunctionClass
 from sympy.utilities.pytest import raises
 
 # make z1 with call-depth = 1
```
Comment (Oktosha) [inline sympy/core/tests/test_var.py]: The comment `# make z1 with call-depth = 1` belonged to the `_make_z1` function but now it's lonely :( Either move it to the function it describes or delete it.

File: sympy/core/tests/test_var.py [hunk 3]
```diff
@@ -1,83 +1,91 @@
...
 
 
-def _make_z1():
-    var("z1")
 
 # make z2 with call-depth = 2
```
Comment (Oktosha) [inline sympy/core/tests/test_var.py]: The same as for the previous comment. The comment `# make z2 with call-depth = 2` belongs to the function `_make_z2` so should be near it.

**Review by Oktosha — CHANGES_REQUESTED** [review]

File: sympy/core/tests/test_var.py (hunk 3 above)
Comment (sachin-4099) [reply sympy/core/tests/test_var.py]: Fine. Will do the necessary changes immediately.

Comment (sachin-4099) [reply sympy/core/tests/test_var.py]: Shall I do it immediately, as then it will restart the TRAVIS CI build?

**Comment by sachin-4099** [comment]
Please review the pull request @oscarbenjamin @Oktosha 
If any changes are still required, please suggest, otherwise I think the request is ready to be merged.

File: sympy/core/tests/test_var.py [hunk 4]
```diff
@@ -1,83 +1,87 @@
 # Tests for var are in their own file, because var pollutes global namespace.
```
Comment (Oktosha) [inline sympy/core/tests/test_var.py]: The comment at the beginning isn't needed anymore as now all the calls to `var` are inside `exec` and don't pollute the global namespace

File: sympy/core/tests/test_var.py [hunk 5]
```diff
@@ -1,83 +1,87 @@
...
+assert fg == Symbol('fg')
 
+# check return value
+assert v == [d, e, fg]
+
+# make z1 with call-depth = 1
 
 def _make_z1():
-    var("z1")
+    z1 = var("z1")
```
Comment (Oktosha) [inline sympy/core/tests/test_var.py]: Why have you added `z1 = var("z1")`? If I'm not mistaken, the whole point was that call `var("z1")` itself creates a variable =/ The same applies to `z2 = var("z2")`

**Review by Oktosha — CHANGES_REQUESTED** [review]

File: sympy/core/tests/test_var.py (hunk 5 above)
Comment (sachin-4099) [reply sympy/core/tests/test_var.py]: Done the required changes and committed them.

**Comment by sachin-4099** [comment]
Please review the pull request @oscarbenjamin @Oktosha
If any changes are still required, please suggest, otherwise I think the request is ready to be merged.

**Comment by Oktosha** [comment]
As discussed in #17547 it's not cool that so much code is inside the strings. Can you make it more granular by splitting the code when possible and using eval for testing for the return values?

**Comment by lkbhitesh07** [comment]
> As discussed in #17547 it's not cool that so much code is inside the strings. Can you make it more granular by splitting the code when possible and using eval for testing for the return values?

@Oktosha i want to work on it but i'm not sure what you are asking to do, will you please elaborate to me maybe via example, so i can start working on my first issue, hope you'll help. I mean how should we use eval, you explained exec to Sachin and i've read that but how to work on this??

**Comment by sachin-4099** [comment]
> As discussed in #17547 it's not cool that so much code is inside the strings. Can you make it more granular by splitting the code when possible and using eval for testing for the return values?

I have not used 'exec', but have used 'eval' and have tried to make everything just as @asmeurer had stated. Please suggest if anything is wrong, I will make the changes ASAP.

**Comment by lkbhitesh07** [comment]
> > As discussed in #17547 it's not cool that so much code is inside the strings. Can you make it more granular by splitting the code when possible and using eval for testing for the return values?
> 
> I have not used 'exec', but have used 'eval' and have tried to make everything just as @asmeurer had stated. Please suggest if anything is wrong, I will make the changes ASAP.

@SACHIN-13 hello, total a newbie in this world, need small help if you are free can you please explain me the syntax you used in eval so that in future it may help
maybe privately on gitter if you are free.

**Comment by sachin-4099** [comment]
> @sachin-4099 hello, total a newbie in this world, need small help if you are free can you please explain me the syntax you used in eval so that in future it may help
> maybe privately on gitter if you are free.

Brother, just refer to the issue #17547 , here at the starting itself @asmeurer has discussed how to do it, and @Oktosha has also given tips on how to proceed, read the conversations. Read the conversations on this page as well to get a more clear idea.

**Comment by sachin-4099** [comment]
Please review the pull request @oscarbenjamin @Oktosha
If any changes are still required, please suggest.

**Comment by oscarbenjamin** [comment]
This looks great. Thanks everyone for the work. Merging

**Pull request merged**